File: ts/packages/knowPro/src/messageIndex.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import {
    generateEmbeddingWithRetry,
    NormalizedEmbedding,
} from "@typeagent/agent-runtime";
import { MessageAccumulator } from "./collections.js";
import { TextEmbeddingIndexSettings } from "./fuzzyIndex.js";
import {
    IMessage,
    MessageOrdinal,
    IndexingEventHandlers,
    TextLocation,
    ListIndexingResult,
    ScoredMessageOrdinal,
    IConversation,
    IMessageTextIndex,
} from "./interfaces.js";
import {
    ITextToTextLocationIndexData,
    ScoredTextLocation,
    TextToTextLocationIndex,
} from "./textLocationIndex.js";
import { getMessageBatches } from "./message.js";

export type MessageTextIndexSettings = {
    embeddingIndexSettings: TextEmbeddingIndexSettings;
    batchSize?: number | number;
};

export interface IMessageTextIndexData {
    indexData?: ITextToTextLocationIndexData | undefined;
    // Number of messages indexed. Absent in data serialized before ordinals
    // were tracked independently of chunk positions.
    messageCount?: number | undefined;
}

export interface IMessageTextEmbeddingIndex extends IMessageTextIndex {
    readonly size: number;
    readonly isEmbeddingEnabled?: boolean;
    generateEmbedding(text: string): Promise<NormalizedEmbedding>;
    lookupByEmbedding(
        textEmbedding: NormalizedEmbedding,
        maxMatches?: number,
        thresholdScore?: number,
        predicate?: (messageOrdinal: MessageOrdinal) => boolean,
    ): ScoredMessageOrdinal[];
    lookupInSubsetByEmbedding(
        textEmbedding: NormalizedEmbedding,
        ordinalsToSearch: MessageOrdinal[],
        maxMatches?: number,
        thresholdScore?: number,
    ): ScoredMessageOrdinal[];
}

export class MessageTextIndex implements IMessageTextEmbeddingIndex {
    public textLocationIndex: TextToTextLocationIndex;
    // Next unassigned message ordinal. Tracked independently of chunk
    // positions because a message can index zero chunks.
    private messageCount: number = 0;

    constructor(public settings: MessageTextIndexSettings) {
        this.textLocationIndex = new TextToTextLocationIndex(
            settings.embeddingIndexSettings,
        );
    }

    /**
     * Number of messages in the index. A message can have several text
     * chunks, and each chunk is one entry in textLocationIndex.
     */
    public get size(): number {
        return this.messageCount;
    }

    /**
     * True only when an embedding model is available. When false, message
     * semantic search is disabled and callers fall back to non-embedding
     * ranking (see isMessageTextEmbeddingIndex and search.ts).
     */
    public get isEmbeddingEnabled(): boolean {
        return (
            this.settings.embeddingIndexSettings.embeddingModel !== undefined
        );
    }

    public addMessages(
        messages: Iterable<IMessage>,
        eventHandler?: IndexingEventHandlers,
    ): Promise<ListIndexingResult> {
        const baseMessageOrdinal: MessageOrdinal = this.size;
        const allChunks: [string, TextLocation][] = [];
        // Collect everything so we can batch efficiently
        let i = 0;
        for (const message of messages) {
            const messageOrdinal = baseMessageOrdinal + i;
            for (
                let chunkOrdinal = 0;
                chunkOrdinal < message.textChunks.length;
                ++chunkOrdinal
            ) {
                allChunks.push([
                    message.textChunks[chunkOrdinal],
                    { messageOrdinal, chunkOrdinal },
                ]);
            }
            ++i;
        }
        this.messageCount = baseMessageOrdinal + i;
        return this.textLocationIndex.addTextLocations(allChunks, eventHandler);
    }

    public async lookupMessages(
        messageText: string,
        maxMatches?: number,
        thresholdScore?: number,
    ): Promise<ScoredMessageOrdinal[]> {
        maxMatches ??= this.settings.embeddingIndexSettings.maxMatches;
        thresholdScore ??= this.settings.embeddingIndexSettings.minScore;
        // Chunks are ranked without a limit so several chunks of one message
        // cannot consume maxMatches; the limit applies per message below.
        const scoredTextLocations = await this.textLocationIndex.lookupText(
            messageText,
            undefined,
            thresholdScore,
        );
        return this.toScoredMessageOrdinals(scoredTextLocations, maxMatches);
    }

    public async lookupMessagesInSubset(
        messageText: string,
        ordinalsToSearch: MessageOrdinal[],
        maxMatches?: number,
        thresholdScore?: number,
    ): Promise<ScoredMessageOrdinal[]> {
        const scoredTextLocations =
            await this.textLocationIndex.lookupTextInSubset(
                messageText,
                this.getChunkPositions(ordinalsToSearch),
                undefined,
                thresholdScore,
            );
        return this.toScoredMessageOrdinals(scoredTextLocations, maxMatches);
    }

    public generateEmbedding(text: string): Promise<NormalizedEmbedding> {
        // Note: if you rename generateEmbedding, be sure to also fix isMessageTextEmbeddingIndex
        const embeddingModel =
            this.settings.embeddingIndexSettings.embeddingModel;
        if (embeddingModel === undefined) {
            throw new Error(
                "Message text embedding index is disabled (no embedding model configured)",
            );
        }
        return generateEmbeddingWithRetry(embeddingModel, text);
    }

    public lookupByEmbedding(
        textEmbedding: NormalizedEmbedding,
        maxMatches?: number,
        thresholdScore?: number,
        predicate?: (messageOrdinal: MessageOrdinal) => boolean,
    ): ScoredMessageOrdinal[] {
        // The text location index passes chunk positions to its predicate
        const chunkPredicate = predicate
            ? (chunkPos: number) =>
                  predicate(this.textLocationIndex.get(chunkPos).messageOrdinal)
            : undefined;
        // Same per-message limit rule as lookupMessages: do not cap chunks.
        const scoredTextLocations = this.textLocationIndex.lookupByEmbedding(
            textEmbedding,
            undefined,
            thresholdScore,
            chunkPredicate,
        );
        return this.toScoredMessageOrdinals(scoredTextLocations, maxMatches);
    }

    public lookupInSubsetByEmbedding(
        textEmbedding: NormalizedEmbedding,
        ordinalsToSearch: MessageOrdinal[],
        maxMatches?: number,
        thresholdScore?: number,
    ): ScoredMessageOrdinal[] {
        const scoredTextLocations =
            this.textLocationIndex.lookupInSubsetByEmbedding(
                textEmbedding,
                this.getChunkPositions(ordinalsToSearch),
                undefined,
                thresholdScore,
            );
        return this.toScoredMessageOrdinals(scoredTextLocations, maxMatches);
    }

    // Subset lookups take message ordinals, but the embedding index is
    // addressed by chunk position. Return every chunk position of those messages.
    private getChunkPositions(messageOrdinals: MessageOrdinal[]): number[] {
        const wanted = new Set(messageOrdinals);
        const positions: number[] = [];
        for (let pos = 0; pos < this.textLocationIndex.size; ++pos) {
            if (wanted.has(this.textLocationIndex.get(pos).messageOrdinal)) {
                positions.push(pos);
            }
        }
        return positions;
    }

    public serialize(): IMessageTextIndexData {
        return {
            indexData: this.textLocationIndex.serialize(),
            messageCount: this.messageCount,
        };
    }

    public deserialize(data: IMessageTextIndexData): void {
        if (data.indexData) {
            this.textLocationIndex.clear();
            this.textLocationIndex.deserialize(data.indexData);
        }
        if (data.messageCount !== undefined) {
            this.messageCount = data.messageCount;
            return;
        }
        // Data written before messageCount existed assigned ordinals from
        // chunk counts, so stored ordinals do not match message positions.
        // Each message's chunks share one ordinal, so rank-compressing the
        // distinct stored ordinals restores message positions. Messages with
        // no chunks left no locations; their positions are unrecoverable.
        this.messageCount = 0;
        let previous: MessageOrdinal = -1;
        for (let i = 0; i < this.textLocationIndex.size; ++i) {
            const location = this.textLocationIndex.get(i);
            if (location.messageOrdinal !== previous) {
                previous = location.messageOrdinal;
                ++this.messageCount;
            }
            location.messageOrdinal = this.messageCount - 1;
        }
    }

    // Since a message has multiple chunks, each of which is indexed individually, we can end up
    // with a message matching multiple times. The message accumulator dedupes those and also
    // supports smoothing the scores if needed
    // Lookups rank every chunk and apply maxMatches here, per message,
    // so several chunks of one message cannot use up maxMatches.
    private toScoredMessageOrdinals(
        scoredLocations: ScoredTextLocation[],
        maxMatches?: number,
    ): ScoredMessageOrdinal[] {
        const messageMatches = new MessageAccumulator();
        messageMatches.addMessagesFromLocations(scoredLocations);
        const scored = messageMatches.toScoredMessageOrdinals();
        return maxMatches ? scored.slice(0, maxMatches) : scored;
    }
}

export async function buildMessageIndex(
    conversation: IConversation,
    settings: MessageTextIndexSettings,
    eventHandler?: IndexingEventHandlers,
    batchSize: number = 8,
): Promise<ListIndexingResult> {
    return addToMessageIndex(
        conversation,
        settings,
        0,
        eventHandler,
        batchSize,
    );
}

export async function addToMessageIndex(
    conversation: IConversation,
    settings: MessageTextIndexSettings,
    startAtOrdinal: MessageOrdinal,
    eventHandler?: IndexingEventHandlers,
    batchSize: number = 8,
): Promise<ListIndexingResult> {
    const result: ListIndexingResult = {
        numberCompleted: 0,
    };
    if (conversation.secondaryIndexes) {
        conversation.secondaryIndexes.messageIndex ??= new MessageTextIndex(
            settings,
        );
        const messageIndex = conversation.secondaryIndexes.messageIndex;
        for (const messageBatch of getMessageBatches(
            conversation,
            startAtOrdinal,
            batchSize,
        )) {
            const batchResult = await messageIndex.addMessages(
                messageBatch.value,
                eventHandler,
            );
            result.numberCompleted += batchResult.numberCompleted;
            result.error = batchResult.error;
            if (result.error) {
                break;
            }
        }
    }
    return result;
}

export function isMessageTextEmbeddingIndex(
    messageIndex: IMessageTextIndex,
): messageIndex is IMessageTextEmbeddingIndex {
    const textIndex: IMessageTextEmbeddingIndex =
        messageIndex as IMessageTextEmbeddingIndex;
    return (
        textIndex.generateEmbedding !== undefined &&
        textIndex.lookupByEmbedding !== undefined &&
        textIndex.lookupInSubsetByEmbedding !== undefined &&
        (textIndex.isEmbeddingEnabled ?? true)
    );
}

File: python/ta/typeagent/knowpro/messageindex.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

from collections.abc import AsyncIterable, Iterable
from typing import Callable

from ..aitools.embeddings import NormalizedEmbedding
from .importing import MessageTextIndexSettings
from .interfaces import (
    IConversation,
    IMessage,
    IMessageTextIndex,
    MessageTextIndexData,
    ITermToSemanticRefIndex,
    ListIndexingResult,
    MessageOrdinal,
    ScoredMessageOrdinal,
    TextLocation,
)
from .textlocindex import ScoredTextLocation, TextToTextLocationIndex


async def build_message_index[
    TMessage: IMessage,
    TTermToSemanticRefIndex: ITermToSemanticRefIndex,
](
    conversation: IConversation[TMessage, TTermToSemanticRefIndex],
    settings: MessageTextIndexSettings,
) -> ListIndexingResult:
    csi = conversation.secondary_indexes
    if csi is None:
        return ListIndexingResult(0)
    if csi.message_index is None:
        csi.message_index = MessageTextIndex(settings)
    messages = conversation.messages
    # Convert collection to list for add_messages
    messages_list = [message async for message in messages]
    return await csi.message_index.add_messages(messages_list)


class IMessageTextEmbeddingIndex(IMessageTextIndex):
    async def generate_embedding(self, text: str) -> NormalizedEmbedding: ...

    def lookup_by_embedding(
        self,
        text_embedding: NormalizedEmbedding,
        max_matches: int | None = None,
        threshold_score: float | None = None,
        predicate: Callable[[MessageOrdinal], bool] | None = None,
    ) -> list[ScoredMessageOrdinal]: ...

    def lookup_in_subset_by_embedding(
        self,
        text_embedding: NormalizedEmbedding,
        ordinals_to_search: list[MessageOrdinal],
        max_matches: int | None = None,
        threshold_score: float | None = None,
    ) -> list[ScoredMessageOrdinal]: ...


class MessageTextIndex(IMessageTextEmbeddingIndex):
    def __init__(self, settings: MessageTextIndexSettings):
        self.settings = settings
        self.text_location_index = TextToTextLocationIndex(
            settings.embedding_index_settings
        )

    async def size(self) -> int:
        return await self.text_location_index.size()

    async def is_empty(self) -> bool:
        return await self.text_location_index.is_empty()

    async def add_messages[TMessage: IMessage](
        self,
        messages: Iterable[TMessage],
    ) -> ListIndexingResult:
        base_message_ordinal: MessageOrdinal = await self.text_location_index.size()
        all_chunks: list[tuple[str, TextLocation]] = []
        # Collect everything so we can batch efficiently.
        for message_ordinal, message in enumerate(messages, base_message_ordinal):
            for chunk_ordinal, chunk in enumerate(message.text_chunks):
                all_chunks.append((chunk, TextLocation(message_ordinal, chunk_ordinal)))
        return await self.text_location_index.add_text_locations(all_chunks)

    async def lookup_messages(
        self,
        message_text: str,
        max_matches: int | None = None,
        threshold_score: float | None = None,
    ) -> list[ScoredMessageOrdinal]:
        max_matches = max_matches or self.settings.embedding_index_settings.max_matches
        threshold_score = (
            threshold_score or self.settings.embedding_index_settings.min_score
        )
        scored_text_locations = await self.text_location_index.lookup_text(
            message_text, max_matches, threshold_score
        )
        return self.to_scored_message_ordinals(scored_text_locations)

    async def lookup_messages_in_subset(
        self,
        message_text: str,
        ordinals_to_search: list[MessageOrdinal],
        max_matches: int | None = None,
        threshold_score: float | None = None,
    ) -> list[ScoredMessageOrdinal]:
        scored_text_locations = await self.text_location_index.lookup_text_in_subset(
            message_text, ordinals_to_search, max_matches, threshold_score
        )
        return self.to_scored_message_ordinals(scored_text_locations)

    async def generate_embedding(self, text: str) -> NormalizedEmbedding:
        # Note: if you rename generate_embedding, be sure to also fix is_message_text_embedding_index.
        # TODO: Retries?
        # TODO: Find a prettier API to get an embedding rather than using _vector_base?
        return await self.text_location_index.generate_embedding(text)

    def lookup_in_subset_by_embedding(
        self,
        text_embedding: NormalizedEmbedding,
        ordinals_to_search: list[MessageOrdinal],
        max_matches: int | None = None,
        threshold_score: float | None = None,
    ) -> list[ScoredMessageOrdinal]:
        scored_text_locations = self.text_location_index.lookup_in_subset_by_embedding(
            text_embedding, ordinals_to_search, max_matches, threshold_score
        )
        return self.to_scored_message_ordinals(scored_text_locations)

    def to_scored_message_ordinals(
        self, scored_locations: list[ScoredTextLocation]
    ) -> list[ScoredMessageOrdinal]:
        matches: dict[MessageOrdinal, ScoredMessageOrdinal] = {}

        for sl in scored_locations:
            value = sl.text_location.message_ordinal
            score = sl.score
            match = matches.get(value)
            if match is None:
                matches[value] = ScoredMessageOrdinal(value, score)
            else:
                match.score = max(score, match.score)

        return [
            ScoredMessageOrdinal(
                match.message_ordinal,
                match.score,
            )
            for match in sorted(
                matches.values(), key=lambda match: match.score, reverse=True
            )
        ]

    def serialize(self) -> MessageTextIndexData:
        return MessageTextIndexData(
            indexData=self.text_location_index.serialize(),
        )

    def deserialize(self, data: MessageTextIndexData) -> None:
        index_data = data.get("indexData")
        if index_data is None:
            return
        self.text_location_index.deserialize(index_data)
